apollo-router/src/spec/schema.rs

```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
+
+            let object_field =
+                |field: &hir::FieldDefinition| (field.name().to_owned(), field.ty().into());
+            let implements_interface = |imp: &hir::ImplementsInterface| imp.interface().to_owned();
+            let object_types: HashMap<_, _> = compiler
+                .db
+                .object_types()
+                .iter()
+                .map(|(name, def)| {
+                    (
+                        name.clone(),
+                        ObjectType {
+                            name: name.clone(),
+                            fields: def
+                                .fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.fields_definition()),
+                                )
+                                .map(object_field)
+                                .collect(),
+                            interfaces: def
+                                .implements_interfaces()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
+                        },
+                    )
+                })
+                .collect();
+
+            let interfaces: HashMap<_, _> = compiler
+                .db
+                .interfaces()
+                .iter()
+                .map(|(name, def)| {
+                    (
+                        name.clone(),
+                        Interface {
+                            name: name.clone(),
+                            fields: def
+                                .fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.fields_definition()),
+                                )
+                                .map(object_field)
+                                .collect(),
+                            interfaces: def
+                                .implements_interfaces()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
```

Comment:
I'm wondering if it would be more readable if we used `From` impl for that kind of code ?